File: solutions/2179-most-beautiful-item-for-each-query/most-beautiful-item-for-each-query.cpp
class Solution {
public:
    vector<int> maximumBeauty(vector<vector<int>>& items, vector<int>& queries) {
        // Sort items by price in ascending order
        sort(items.begin(), items.end());
        map<int, int> priceToBeauty;
        int maxBeauty = 0;
        
        // Build price to maximum beauty mapping
        for (const auto& item : items) {
            maxBeauty = max(maxBeauty, item[1]);
            priceToBeauty[item[0]] = maxBeauty;
        }
        
        // Process each query using binary search
        vector<int> result(queries.size());
        for (int i = 0; i < queries.size(); i++) {
            auto it = priceToBeauty.upper_bound(queries[i]);
            result[i] = it == priceToBeauty.begin() ? 0 : prev(it)->second;
        }
        return result;
    }
};

File: solutions/2179-most-beautiful-item-for-each-query/most-beautiful-item-for-each-query.rs
impl Solution {
    pub fn maximum_beauty(mut items: Vec<Vec<i32>>, queries: Vec<i32>) -> Vec<i32> {
        // Step 1: Sort items by price in ascending order
        items.sort_by_key(|item| item[0]);
        
        // Step 2: Extract prices and beauties into separate vectors
        let prices: Vec<i32> = items.iter().map(|item| item[0]).collect();
        let beauties: Vec<i32> = items.iter().map(|item| item[1]).collect();
        
        // Step 3: Create running maximum beauty array
        let mut max_beauties = vec![0];
        let mut current_max = 0;
        for &beauty in beauties.iter() {
            current_max = current_max.max(beauty);
            max_beauties.push(current_max);
        }
        
        // Step 4: Process each query using binary search
        queries.iter().map(|&query| {
            let index = prices.partition_point(|&x| x <= query);
            max_beauties[index]
        }).collect()
    }
}